Develop

Lev2098 opened the pull request.

Lev2098 committed:
add order and ticket serializer

Lev2098 committed:
delete unused

mateacademy-ai-mentor commented on cinema/models.py:
```diff
@@ -84,23 +84,45 @@ class Ticket(models.Model):
-                )
+
+    def clean(self):
+        Ticket.validate_data(
+            self.seat,
+            self.row,
+            self.movie_session.cinema_hall.seats_in_row,
+            self.movie_session.cinema_hall.rows,
+            ValueError
```
The parameters passed to `Ticket.validate_data` are in the wrong order. The correct order should be `self.seat`, `self.movie_session.cinema_hall.seats_in_row`, `self.row`, `self.movie_session.cinema_hall.rows`, `ValidationError`.

mateacademy-ai-mentor commented on cinema/models.py:
```diff
@@ -84,23 +84,45 @@
+            self.row,
+            self.movie_session.cinema_hall.seats_in_row,
+            self.movie_session.cinema_hall.rows,
+            ValueError
```
Instead of raising `ValueError`, you should raise `ValidationError` to properly integrate with Django's validation system.

mateacademy-ai-mentor commented on cinema/serializers.py:
```diff
@@ -78,3 +80,31 @@ class MovieSessionDetailSerializer(MovieSessionSerializer):
+        fields = ("id", "movie_session","row", "seat")
+
+    def validate(self, attrs):
+        Ticket.validate_data(
+            attrs["seat"],
+            attrs["row"],
+            attrs["movie_session"].cinema_hall.seats_in_row,
+            attrs["movie_session"].cinema_hall.rows,
+            serializers.ValidationError
```
The parameters passed to `Ticket.validate_data` are in the wrong order. The correct order should be `attrs['seat']`, `attrs['movie_session'].cinema_hall.seats_in_row`, `attrs['row']`, `attrs['movie_session'].cinema_hall.rows`, `serializers.ValidationError`.

mateacademy-ai-mentor commented on cinema/serializers.py:
```diff
@@ -78,3 +80,31 @@
+    class Meta:
+        model = Order
+        fields = ("id", "created_at", "tickets")
+    with transaction.atomic():
```
The `with transaction.atomic():` block is incorrectly used here. The `create` method should not be indented inside this block. Instead, the transaction should be managed within the method itself.

mateacademy-ai-mentor commented on cinema/views.py:
```diff
@@ -56,3 +56,14 @@ def get_serializer_class(self):
+    serializer_class = OrderSerializer
+
+    def get_queryset(self):
+        return self.queryset.filter(user=self.request.user)
```
In the `get_queryset` method, it's better to directly use `Order.objects.filter(user=self.request.user)` instead of `self.queryset.filter(user=self.request.user)` to avoid potential issues with queryset caching or modifications.

mateacademy-ai-mentor reviewed (changes requested):
Thank you for your hard work on this project! While you've made great progress, there are a few critical issues that need to be addressed to ensure the functionality and security of the application:

1. **Pagination and Filtering**: The task requires pagination for the `GET api/cinema/orders/` endpoint and filtering for movies and movie sessions, which are not currently implemented. These are core requirements that need to be addressed.

2. **Validation and Error Handling**: There are issues with the parameter order in the `Ticket.validate_data` method and the use of `ValueError` instead of `ValidationError`. These need to be corrected to ensure proper validation and error handling.

3. **Transaction Management**: The `create` method in `OrderSerializer` is incorrectly indented within a `with transaction.atomic()` block. This needs to be fixed to ensure proper transaction management.

4. **Security Concerns**: The `SECRET_KEY` is hardcoded, and `DEBUG` is set to `True`, which poses security risks. These should be managed through environment variables.

Please address these issues to meet the task requirements fully. Keep up the great work, and don't hesitate to reach out if you need any help! 😊

✨ Please rate this review in the Platform's notification center. I will be grateful for your feedback! ✨

🔄 Fix the issues, commit the changes, and then re-request my review using the `Re-request review` button in the Reviewers tab.

Lev2098 committed:
Add filtering capabilities to Movie and MovieSession APIs

Integrated django_filters to enable filtering by genres, actors, and title in MovieViewSet and by date and movie in MovieSessionViewSet. Updated serializers and models to support these filter operations, ensuring more flexible and comprehensive data retrieval. Also updated requirements.txt to include the django-filter dependency.

mateacademy-ai-mentor reviewed (commented):
Sorry, the pull request review was unsuccessful for the following reasons:

> [!WARNING]
> **There are no files to review.** Please make sure that you have added files to the pull request. Some files and directories may be ignored from the review or failed to load.
> <sub>*[CODE: 4]*</sub>

**Additional info:**
*Ignored files list:*
- `cinema/models.py`
- `cinema/serializers.py`
- `cinema/tests/test_order_api.py`
- `cinema/urls.py`
- `cinema/views.py`
- `cinema_service/settings.py`
- `requirements.txt`

If the issue is within your control, please address it.
The next review attempt <ins>will automatically launch</ins> in 1 hour.
If you believe this is an error or need assistance, please contact the support team.

Lev2098 committed:
django 4.2

Lev2098 committed:
django 4.1

Lev2098 committed:
django-filter~=23.5

Lev2098 committed:
Add pagination to OrderViewSet with large result set

Lev2098 committed:
fix flake8

mateacademy-ai-mentor reviewed (commented):
Sorry, the pull request review was unsuccessful for the following reasons:

> [!WARNING]
> **There are no files to review.** Please make sure that you have added files to the pull request. Some files and directories may be ignored from the review or failed to load.
> <sub>*[CODE: 4]*</sub>

**Additional info:**
*Ignored files list:*
- `cinema/models.py`
- `cinema/pagination.py`
- `cinema/serializers.py`
- `cinema/tests/test_order_api.py`
- `cinema/urls.py`
- `cinema/views.py`
- `cinema_service/settings.py`
- `requirements.txt`

If the issue is within your control, please address it.
The next review attempt <ins>will automatically launch</ins> in 1 hour.
If you believe this is an error or need assistance, please contact the support team.

mateacademy-ai-mentor reviewed (commented):
Sorry, the pull request review was unsuccessful for the following reasons:

> [!WARNING]
> **There are no files to review.** Please make sure that you have added files to the pull request. Some files and directories may be ignored from the review or failed to load.
> <sub>*[CODE: 4]*</sub>

**Additional info:**
*Ignored files list:*
- `cinema/models.py`
- `cinema/pagination.py`
- `cinema/serializers.py`
- `cinema/tests/test_order_api.py`
- `cinema/urls.py`
- `cinema/views.py`
- `cinema_service/settings.py`
- `requirements.txt`

If the issue is within your control, please address it.
The next review attempt <ins>will automatically launch</ins> in 1 hour.
If you believe this is an error or need assistance, please contact the support team.

mateacademy-ai-mentor reviewed (commented):
Sorry, the pull request review was unsuccessful for the following reasons:

> [!WARNING]
> **There are no files to review.** Please make sure that you have added files to the pull request. Some files and directories may be ignored from the review or failed to load.
> <sub>*[CODE: 4]*</sub>

**Additional info:**
*Ignored files list:*
- `cinema/models.py`
- `cinema/pagination.py`
- `cinema/serializers.py`
- `cinema/tests/test_order_api.py`
- `cinema/urls.py`
- `cinema/views.py`
- `cinema_service/settings.py`
- `requirements.txt`

If the issue is within your control, please address it.
The next review attempt <ins>will automatically launch</ins> in 1 hour.
If you believe this is an error or need assistance, please contact the support team.

pavlopro commented on cinema/models.py:
```diff
@@ -79,28 +83,50 @@ class Ticket(models.Model):
+            self.movie_session.cinema_hall.rows,
+            ValidationError,
+        )
+        # for ticket_attr_value, ticket_attr_name, cinema_hall_attr_name in [
```
please remove comments

pavlopro commented on cinema/tests/test_order_api.py:
```diff
@@ -58,6 +58,7 @@ def test_get_order(self):
         orders_response = self.client.get("/api/cinema/orders/")
         self.assertEqual(orders_response.status_code, status.HTTP_200_OK)
         self.assertEqual(orders_response.data["count"], 1)
+        print("Response data:", orders_response.data)
```
please remove this from test file

pavlopro commented on cinema_service/settings.py:
```diff
@@ -136,3 +137,9 @@
 
 DEFAULT_AUTO_FIELD = "django.db.models.BigAutoField"
+
+# REST_FRAMEWORK = {
```
please remove commented code

Lev2098 committed:
Remove commented-out code and debug print statements

vsmutok commented on cinema/views.py:
```diff
@@ -11,7 +13,7 @@
     MovieDetailSerializer,
     MovieSessionDetailSerializer,
-    MovieListSerializer,
+    MovieListSerializer, OrderSerializer,
```
```suggestion
    MovieListSerializer, 
    OrderSerializer,
```

vsmutok commented on requirements.txt:
```diff
@@ -4,4 +4,5 @@ flake8-quotes==3.3.1
-djangorestframework==3.13.1
\ No newline at end of file
+djangorestframework==3.13.1
+django-filter==23.5
```
add new empty line

vsmutok commented on cinema/urls.py:
```diff
@@ -6,7 +6,7 @@
     CinemaHallViewSet,
     MovieViewSet,
-    MovieSessionViewSet,
+    MovieSessionViewSet, OrderViewSet,
```
```suggestion
    MovieSessionViewSet, 
    OrderViewSet,
```

Lev2098 committed:
add new empty line

Lev2098 committed:
Separate import statements for readability.

vsmutok reviewed (approved):
Great!
